### PR TITLE
【Feature】【NPU】KV Cache Optimization for 1/TP Caching in the Prefill Phase

### PR DESCRIPTION
**Pull request by code0-ling**
## Motivation

For the MLA model, TP sharding leads to the replication of the KV Cache, which wastes video memory space; enable_kv_storage_optimization_mla sharding reduces such waste by partitioning the stored KV Cache to save video memory space.
Overall logic for the PD-separated scenario:
1、In the prefill phase, video memory space for the KV Cache of the 1/TP sequence is allocated and stored;
2、No adjustment is made in the decode phase, where video memory space for the KV Cache of the full sequence is allocated instead;
3、Each rank on the D node needs to establish a connection with every rank on the P node and obtain the corresponding KV Cache from them.

## Accuracy Tests

Single-node TP2+TP4：Use the DeepSeek-V2-Lite-W8A8 model for testing.
image

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

---
### CI States

Latest PR Test (Base): :x: **Missing `run-ci` label** -- add it to run CI tests.
Latest PR Test (Extra): :x: **Blocked** -- `run-ci` is required first.

### CODE REVIEW

File: python/sglang/srt/disaggregation/common/conn.py [hunk 1]
```diff
@@ -370,6 +370,27 @@ def __init__(
                     self.prefill_attn_tp_size // self.kv_mgr.attn_tp_size
                 ) * (self.prefill_pp_size // self.kv_mgr.pp_size)
 
+        self.prefill_kv_split_size = 1
```
Comment (husf1130) [inline python/sglang/srt/disaggregation/common/conn.py:768]: recommended to directly reuse `prefill_attn_tp_size`

File: python/sglang/srt/disaggregation/prefill.py [hunk 2]
```diff
@@ -315,6 +299,16 @@ def pop_bootstrapped(
             assert req.metadata_buffer_index is not None
 
             num_pages = kv_to_page_num(num_kv_indices, self.token_to_kv_pool.page_size)
+            if get_global_server_args().enable_kv_storage_optimization_mla:
+                size = get_attention_tp_size()
+                rank = get_attention_tp_rank()
```
Comment (husf1130) [inline python/sglang/srt/disaggregation/prefill.py]: Suggest naming them `tp_rank` and `tp_size`

File: python/sglang/srt/disaggregation/prefill.py [hunk 3]
```diff
@@ -315,6 +299,16 @@ def pop_bootstrapped(
             assert req.metadata_buffer_index is not None
 
             num_pages = kv_to_page_num(num_kv_indices, self.token_to_kv_pool.page_size)
```
Comment (husf1130) [inline python/sglang/srt/disaggregation/prefill.py]: This line should be moved into the else block.

File: python/sglang/srt/disaggregation/common/conn.py [hunk 4]
```diff
@@ -370,6 +370,27 @@ def __init__(
...
+        if get_global_server_args().enable_kv_storage_optimization_mla:
+            self.prefill_kv_split_size = self.prefill_attn_tp_size
+            # tp_rank in prefill needs send_kvcache to all tp_rank in tp_group in decode
+            self.required_dst_info_num = self.kv_mgr.attn_tp_size
+
+            # All tp_rank in tp_group need to be notified
+            self.target_tp_ranks = [rank for rank in range(self.prefill_attn_tp_size)]
```
Comment (husf1130) [inline python/sglang/srt/disaggregation/common/conn.py]: The number of elements in the target_tp_ranks should be equal to `required_prefill_response_num`.

File: python/sglang/srt/disaggregation/common/conn.py [hunk 5]
```diff
@@ -370,6 +370,27 @@ def __init__(
...
+            page_size = self.kv_mgr.kv_args.page_size
+            total_page_num = (
+                self.kv_mgr.kv_args.origin_input_len + page_size - 1
+            ) // page_size
+            base_page = total_page_num // self.prefill_kv_split_size
+            extra = total_page_num % self.prefill_kv_split_size
+            self.required_prefill_response_num = (
+                self.prefill_kv_split_size if base_page >= 1 else extra
+            )
+
```
Comment (husf1130) [inline python/sglang/srt/disaggregation/common/conn.py:785]: When enable_kv_storage_optimization_mla is enabled, the calculation logic here should be mutually exclusive with the above if-elif-else statement, instead of being recalculated and overwritten.

File: python/sglang/srt/utils/common.py [hunk 6]
```diff
@@ -3574,6 +3562,21 @@ def placeholder(*args, **kwargs):
     return final_module, None
 
 
+# Calculate the page range alloced to split_kv_rank
+def get_split_kv_page_range(split_kv_size, split_kv_rank, total_page_num):
```
Comment (husf1130) [inline python/sglang/srt/utils/common.py]: suggest naming the first parameter `split_kv_rank_size`.

File: python/sglang/srt/server_args.py [hunk 7]
```diff
@@ -2432,6 +2429,19 @@ def _handle_load_format(self):
...
 
+        if self.tp_size == 1 and self.enable_kv_storage_optimization_mla:
+            logger.warning(
+                "enable_kv_storage_optimization_mla is adjust tp False when tp_size=1"
+            )
+            self.enable_kv_storage_optimization_mla = False
+        if (
+            self.enable_kv_storage_optimization_mla
+            and self.disaggregation_mode == "null"
+        ):
```
Comment (husf1130) [inline python/sglang/srt/server_args.py]: Append checks for `non-MLA attention` and `pipeline parallelism`.

File: python/sglang/srt/server_args.py [hunk 8]
```diff
@@ -4949,6 +4959,13 @@ def add_cli_args(parser: argparse.ArgumentParser):
             help="JSON-formatted forward hook specifications to attach to the model.",
         )
 
+        # For optimization kv storage in pd-disaggregation (only prefill)
+        parser.add_argument(
+            "--enable-kv-storage-optimization-mla",
+            action="store_true",
+            help="Use optimization kv storage in pd-disaggregation (only prefill) with set --disable-radix-cache and --attention-backend ascend and MLA",
```
Comment (husf1130) [inline python/sglang/srt/server_args.py]: This feature should not exclude other hardware environments.

File: python/sglang/srt/server_args.py [hunk 9]
```diff
@@ -4949,6 +4959,13 @@ def add_cli_args(parser: argparse.ArgumentParser):
             help="JSON-formatted forward hook specifications to attach to the model.",
         )
 
+        # For optimization kv storage in pd-disaggregation (only prefill)
```
Comment (husf1130) [inline python/sglang/srt/server_args.py]: Is this parameter only used in prefill instance?

File: python/sglang/srt/mem_cache/common.py [hunk 10]
```diff
@@ -121,7 +132,19 @@ def write_cache_indices(
                 (req_idx, slice(prefix_len, seq_len)),
                 out_cache_loc[pt : pt + extend_len],
             )
-            pt += extend_len
+            if get_global_server_args().enable_kv_storage_optimization_mla:
```
Comment (husf1130) [inline python/sglang/srt/mem_cache/common.py:157]: Double writes should be avoided.

File: python/sglang/srt/mem_cache/common.py [hunk 11]
```diff
@@ -325,6 +348,81 @@ def alloc_req_slots(
...
+        prefix_lens=prefix_lens,
+        prefix_lens_cpu=prefix_lens_cpu,
+        seq_lens=token_num_tensor,
+        seq_lens_cpu=seq_lens_cpu,
+        last_loc=last_loc,
+        extend_num_tokens=extend_num_token,
+    )
+
+    # padding -1 to origin input lenth, skip this loc-1 when set_kv_buffer
+    len_sum = sum(l.item() for l in seq_lens)
```
Comment (husf1130) [inline python/sglang/srt/mem_cache/common.py]: replace with `len_sum = seq_lens.sum().item()`

File: python/sglang/srt/mem_cache/common.py [hunk 12]
```diff
@@ -325,6 +348,81 @@ def alloc_req_slots(
...
+        tp_seq_start[i] = start_page * page_size
+
+    # alloc out_loc use tokens length in split_kv_rank
+    extend_num_token = sum(l.item() for l in token_num_tensor)
+    tp_out_loc = alloc_paged_token_slots_extend(
+        tree_cache=batch.tree_cache,
+        prefix_lens=prefix_lens,
+        prefix_lens_cpu=prefix_lens_cpu,
+        seq_lens=token_num_tensor,
+        seq_lens_cpu=seq_lens_cpu,
```
Comment (husf1130) [inline python/sglang/srt/mem_cache/common.py:494]: Maybe this parameter `seq_lens_cpu` should be equal to `torch.tensor([<individual_token_numbers>], dtype=torch.int64)`.

File: python/sglang/srt/mem_cache/common.py [hunk 13]
```diff
@@ -325,6 +348,81 @@ def alloc_req_slots(
...
+        if end_page == page_num - 1:
+            token_num_tensor[i] = (page_count - 1) * page_size + last_page_token_num
+        else:
+            token_num_tensor[i] = page_count * page_size
+
+        # For short prompt, exeit empty page
+        token_num_tensor[i] = max(token_num_tensor[i], 0)
+        batch.reqs[i].tp_seq_len = token_num_tensor[i]
+
+        tp_seq_start[i] = start_page * page_size
```
Comment (husf1130) [inline python/sglang/srt/mem_cache/common.py]: replace with `tp_token_index_start_in_seq`

File: python/sglang/srt/mem_cache/common.py [hunk 14]
```diff
@@ -325,6 +348,81 @@ def alloc_req_slots(
     return req_pool_indices
 
 
+# For PD disaggregation prefill,aligned by page
+def alloc_tp_paged_token_slots_extend(
```
Comment (husf1130) [inline python/sglang/srt/mem_cache/common.py]: replace with `alloc_paged_token_slots_extend_for_mla_kvcache_split`

File: python/sglang/srt/mem_cache/chunk_cache.py [hunk 15]
```diff
@@ -68,12 +79,14 @@ def cache_finished_req(self, req: Req, is_insert: bool = True):
         kv_indices = self.req_to_token_pool.req_to_token[
```
Comment (husf1130) [inline python/sglang/srt/mem_cache/chunk_cache.py]: remove this line

File: python/sglang/srt/mem_cache/chunk_cache.py [hunk 16]
```diff
@@ -58,6 +59,16 @@ def match_prefix(self, params: MatchPrefixParams) -> MatchResult:
...
         )
 
+    def process_split_kv_indices(self, req: Req, kv_indices):
+        if (
+            get_global_server_args().enable_kv_storage_optimization_mla
+            and get_global_server_args().disaggregation_mode == "prefill"
+        ):
+            kv_indices = self.req_to_token_pool.req_to_token[
+                req.req_pool_idx, : req.tp_seq_len
+            ]
```
Comment (husf1130) [inline python/sglang/srt/mem_cache/chunk_cache.py:83]: add
`else:
     kv_indices = self.req_to_token_pool.req_to_token[
         req.req_pool_idx, : ids_len
     ]
`

`ids_len` is the 3rd parameter

File: python/sglang/srt/mem_cache/chunk_cache.py [hunk 17]
```diff
@@ -68,12 +79,14 @@ def cache_finished_req(self, req: Req, is_insert: bool = True):
         kv_indices = self.req_to_token_pool.req_to_token[
             req.req_pool_idx, :kv_committed_len
         ]
+        kv_indices = self.process_split_kv_indices(req, kv_indices)
         self.token_to_kv_pool_allocator.free(kv_indices)
 
     def cache_unfinished_req(self, req: Req, chunked=False):
         kv_indices = self.req_to_token_pool.req_to_token[
```
Comment (husf1130) [inline python/sglang/srt/mem_cache/chunk_cache.py]: remove this line

File: python/sglang/srt/disaggregation/common/conn.py (hunk 1 above)
Comment (code0-ling) [reply python/sglang/srt/disaggregation/common/conn.py:768]: Modified already.

File: python/sglang/srt/disaggregation/prefill.py (hunk 3 above)
Comment (code0-ling) [reply python/sglang/srt/disaggregation/prefill.py]: Modified already.

File: python/sglang/srt/disaggregation/common/conn.py (hunk 4 above)
Comment (code0-ling) [reply python/sglang/srt/disaggregation/common/conn.py]: target_tp_ranks indicates the actual number of cards used for prefill. required_prefill_response_num should be equal to target_tp_ranks, instead of target_tp_ranks being equal to required_prefill_response_num.

File: python/sglang/srt/disaggregation/common/conn.py (hunk 5 above)
Comment (code0-ling) [reply python/sglang/srt/disaggregation/common/conn.py:785]: Modified already.

File: python/sglang/srt/mem_cache/common.py (hunk 14 above)
Comment (code0-ling) [reply python/sglang/srt/mem_cache/common.py]: Modified already.

File: python/sglang/srt/mem_cache/common.py [hunk 18]
```diff
@@ -112,16 +122,30 @@ def write_cache_indices(
...
+                    (req_idx, slice(prefix_len, seq_lens_tp)),
+                    tp_out_loc[pt:pt + extend_lens_tp]
+                )
+                pt += extend_lens_tp
+            else:
+                req_to_token_pool.write(
+                    (req_idx, slice(prefix_len, seq_len)),
+                    out_cache_loc[pt:pt + extend_len],
+                )
+                pt += extend_len
```
[the remark this answers is not in the document]
Comment (lfy-code123) [reply python/sglang/srt/mem_cache/common.py]: Modified already.

File: python/sglang/srt/mem_cache/common.py (hunk 10 above)
Comment (lfy-code123) [reply python/sglang/srt/mem_cache/common.py]: Modified already.

File: python/sglang/srt/utils/common.py (hunk 6 above)
Comment (lfy-code123) [reply python/sglang/srt/utils/common.py]: Modified already.
